File: frontend/src/lib/schemas.ts
import { z } from "zod";

export const LoginSchema = z.object({
  email: z.string().email("Invalid email address"),
  password: z.string().min(6, "Password must be at least 6 characters"),
});

export const RegisterSchema = z.object({
  name: z.string().min(2, "Name must be at least 2 characters"),
  email: z.string().email("Invalid email address"),
  password: z.string().min(6, "Password must be at least 6 characters"),
});

export const CreateInterviewSchema = z.object({
  candidate_email: z.string().email("Invalid email address"),
  role: z.string().min(1, "Role is required"),
  department: z.string().optional(),
  type: z.enum(["TECHNICAL", "HR", "MANAGERIAL", "CULTURE_FIT", "FINAL_ROUND"]),
  mode: z.enum(["ONLINE", "IN_PERSON", "PHONE"]),
  date: z.string().refine((val) => {
    const d = new Date(val);
    return !isNaN(d.getTime()) && d > new Date();
  }, {
    message: "Interview date must be in the future",
  }),
  duration_min: z.preprocess(
    (val) => parseInt(val as string, 10),
    z.number().min(15, "Duration must be at least 15 mins").max(240, "Duration cannot exceed 240 mins")
  ),
  interviewer: z.string().optional(),
  meeting_link: z.string().url("Invalid URL").or(z.literal("")).optional(),
  notes: z.string().max(1000, "Notes cannot exceed 1000 characters").optional(),
});

export const CreateNoticeSchema = z.object({
  title: z.string().min(5, "Title must be at least 5 characters").max(200),
  body: z.string().min(10, "Body must be at least 10 characters").max(5000),
  type: z.enum(["GENERAL", "REMINDER", "UPDATE", "IMPORTANT", "HOLIDAY"]),
  priority: z.enum(["LOW", "MEDIUM", "HIGH"]),
});

File: backend/shared/schemas.py
from pydantic import BaseModel, EmailStr, HttpUrl, field_validator
from typing import Literal, Optional
from datetime import datetime, timezone

class CreateInterviewSchema(BaseModel):
    candidate_email: EmailStr
    role:            str
    department:      Optional[str] = None
    type:            Literal["TECHNICAL", "HR", "MANAGERIAL", "CULTURE_FIT", "FINAL_ROUND"]
    mode:            Literal["ONLINE", "IN_PERSON", "PHONE"]
    date:            datetime
    duration_min:    int = 60
    interviewer:     Optional[str] = None
    meeting_link:    Optional[HttpUrl] = None
    notes:           Optional[str] = None
    category:        Optional[str] = ""

    @field_validator("duration_min")
    def check_duration(cls, v):
        if not (15 <= v <= 240):
            raise ValueError("duration_min must be 15–240")
        return v

    @field_validator("meeting_link", mode="before")
    @classmethod
    def empty_string_to_none(cls, v):
        if v == "":
            return None
        return v

    @field_validator("date")
    def must_be_future(cls, v):
        # Handle timezone-aware or naive datetimes safely
        now = datetime.now(timezone.utc) if v.tzinfo else datetime.now()
        if v <= now:
            raise ValueError("Interview date must be in the future")
        return v

class UpdateStatusSchema(BaseModel):
    status: Literal["PENDING", "CONFIRMED", "COMPLETED", "CANCELLED", "RESCHEDULED"]

class CreateNoticeSchema(BaseModel):
    title:    str
    body:     str
    type:     Literal["GENERAL", "REMINDER", "UPDATE", "IMPORTANT", "HOLIDAY"]
    priority: Literal["LOW", "MEDIUM", "HIGH"]
